pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp

```diff
@@ -359,33 +360,128 @@ class StateVectorCudaManaged
+                std::vector<CFP_t> matrix_cu(gate_matrix.size());
+                std::transform(
+                    gate_matrix.begin(), gate_matrix.end(), matrix_cu.begin(),
+                    [](const std::complex<Precision> &x) {
+                        return cuUtil::complexToCu<std::complex<Precision>>(x);
+                    });
+
+                gate_cache_.add_gate(opName, par[0], matrix_cu);
+            }
+            std::reverse(ctrlsInt.begin(), ctrlsInt.end());
+            std::reverse(tgtsInt.begin(), tgtsInt.end());
+            std::reverse(ctrls_valuesInt.begin(), ctrls_valuesInt.end());
+            applyDeviceGeneralGate_(
+                gate_cache_.get_gate_device_ptr(opName, par[0]), ctrlsInt,
+                tgtsInt, ctrls_valuesInt, adjoint);
         }
     }
 
+    /**
+     * @brief Apply a single gate to the state vector.
+     *
+     * @param gate_matrix Gate matrix data (in row-major format).
+     * @param matrix_size Size of the matrix.
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (false or true).
+     * @param tgt_wires Target wires to apply gate to.
+     * @param inverse Indicates whether to use adjoint of gate.
+     */
+    void applyControlledMatrix(const ComplexT *gate_matrix,
+                               const std::size_t matrix_size,
+                               const std::vector<std::size_t> &controlled_wires,
+                               const std::vector<bool> &controlled_values,
+                               const std::vector<std::size_t> &tgt_wires,
+                               bool inverse = false) {
+        PL_ABORT_IF_NOT(
+            areVecsDisjoint<std::size_t>(controlled_wires, tgt_wires),
+            "`controlled_wires` and `target wires` must be disjoint.");
+        PL_ABORT_IF(controlled_wires.size() != controlled_values.size(),
+                    "`controlled_wires` and `controlled_values` must have the "
+                    "same size.");
+        DataBuffer<CFP_t, int> d_matrix{
+            matrix_size, BaseType::getDataBuffer().getDevTag(), true};
+        d_matrix.CopyHostDataToGpu(gate_matrix, matrix_size, false);
+        // ensure wire indexing correctly preserved for tensor-observables
+        auto ctrlsInt = NormalizeCastIndices<std::size_t, int>(
+            controlled_wires, BaseType::getNumQubits());
+        auto tgtsInt = NormalizeCastIndices<std::size_t, int>(
+            tgt_wires, BaseType::getNumQubits());
+        auto ctrls_valuesInt =
+            Pennylane::Util::cast_vector<bool, int>(controlled_values);
+
+        std::reverse(ctrlsInt.begin(), ctrlsInt.end());
+        std::reverse(tgtsInt.begin(), tgtsInt.end());
+        std::reverse(ctrls_valuesInt.begin(), ctrls_valuesInt.end());
+
+        applyDeviceGeneralGate_(d_matrix.getData(), ctrlsInt, tgtsInt,
+                                ctrls_valuesInt, inverse);
```

Comment:
Could we hack the reverse calls in the implementation of `applyDeviceGeneralGate_` instead? This may increase the code readability.

Please add code comments about the necessity of reversing the order of wires for LGPU.

Reply:
Good call! These line will be moved to applyDeviceGeneralGate_ in PR#959 after unifying memory layout!